src/backend/optimizer/util/pathnode.c

```diff
@@ -2401,17 +2404,23 @@ create_valuesscan_path(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 	pathnode->parent = rel;
 	pathnode->pathkeys = NIL;	/* result is always unordered */
 
-    /*
-     * CDB: If VALUES list contains mutable functions, evaluate it on entry db.
-     * Otherwise let it be evaluated in the same slice as its parent operator.
-     */
-    Assert(rte->rtekind == RTE_VALUES);
-    if (contain_mutable_functions((Node *)rte->values_lists))
-        CdbPathLocus_MakeEntry(&pathnode->locus);
-    else
-        CdbPathLocus_MakeGeneral(&pathnode->locus);
+	/*
+	 * CDB: If VALUES list contains mutable functions, evaluate it on entry db.
+	 * If volatile function is contained in predicate, evaluate as singleton.
+	 * Otherwise let it be evaluated in the same slice as its parent operator.
+	 *
+	 * TODO: if values list contains parameter from external scan, locustype
+	 * have to be inherited from that scan
+	 */
+	Assert(rte->rtekind == RTE_VALUES);
+	if (contain_mutable_functions((Node *)rte->values_lists))
+		CdbPathLocus_MakeEntry(&pathnode->locus);
+	else if (contain_volatile_functions((Node *) rel->baserestrictinfo))
```

Comment:
The interesting part that there are no such changes in this and create_functionscan_path() functions in 6x. Does it mean we should fix something in 6X too?